### PR TITLE
Adding net support to docker create

### LINKED ISSUE #111: Create network scope
This is the creation of a network scope, which has associated:
* name
* IPAM range and management
* Distributed Port Group, or
* vSwitch
* ...

When determining the backing DPG, this should be done each time the appliance initializes it's state and, if necessary created. This means there is a natural update when we introduce programmatic creation of NSX logical networks

### PR DESCRIPTION
**Pull request by hmahmood**
Fixes #111 #107 

Adding net support to docker create
Adding net support to port layer exec create

### CODE REVIEW

File: apiservers/engine/backends/container.go [hunk 1]
```diff
@@ -255,6 +255,10 @@ func (c *Container) ContainerAttach(name string, cac *backend.ContainerAttachCon
 // Utility Functions
 //----------
 
+func strPtr(s string) *string {
```
Comment (fdawg4l) [inline apiservers/engine/backends/container.go]: Doesn't look like this is called anywhere.

File: apiservers/engine/backends/container.go [hunk 2]
```diff
@@ -284,6 +288,21 @@ func (c *Container) dockerContainerCreateParamsToPortlayer(cc types.ContainerCre
...
+		}
+		if cc.NetworkingConfig != nil {
+			if es, ok := cc.NetworkingConfig.EndpointsConfig[nc.NetworkName]; ok {
+				if es.IPAMConfig != nil {
+					nc.Address = &es.IPAMConfig.IPv4Address
+				}
+			}
+		}
+
+		portLayerConfig.CreateConfig.NetworkSettings = nc
```
Comment (fdawg4l) [inline apiservers/engine/backends/container.go]: Consider moving everything after the `if` to a helper.  Then we can test this path without bringing up the full daemon.

File: apiservers/portlayer/restapi/handlers/exec_handlers.go [hunk 3]
```diff
@@ -75,10 +77,67 @@ func (handler *ExecHandlersImpl) Configure(api *operations.PortLayerAPI) {
...
 
+func addContainerToScope(name string, ns *models.NetworkConfig) (*metadata.NetworkEndpoint, *network.Scope, error) {
+	if ns == nil {
+		return nil, nil, nil
+	}
+
+	var err error
+	var s *network.Scope
+	switch ns.NetworkName {
+	case "default":
```
Comment (fdawg4l) [inline apiservers/portlayer/restapi/handlers/exec_handlers.go]: If this is "default", why `default:` below.  A comment is needed at least.

File: apiservers/portlayer/restapi/handlers/exec_handlers.go [hunk 4]
```diff
@@ -75,10 +77,67 @@ func (handler *ExecHandlersImpl) Configure(api *operations.PortLayerAPI) {
...
+		if err != nil || len(scopes) != 1 {
+			return nil, nil, err
+		}
+
+		s = scopes[0]
+	}
+
+	var ip *net.IP
+	if ns.Address != nil {
+		i := net.ParseIP(*ns.Address)
```
Comment (fdawg4l) [inline apiservers/portlayer/restapi/handlers/exec_handlers.go]: Just checking. This type assertion will never fail, right?

File: apiservers/portlayer/restapi/handlers/exec_handlers.go [hunk 5]
```diff
@@ -75,10 +77,67 @@ func (handler *ExecHandlersImpl) Configure(api *operations.PortLayerAPI) {
...
+		s = GetNetworkContext().DefaultScope()
+
+	default:
+		var scopes []*network.Scope
+		scopes, err = GetNetworkContext().Scopes(&ns.NetworkName)
+		if err != nil || len(scopes) != 1 {
+			return nil, nil, err
+		}
+
+		s = scopes[0]
```
Comment (fdawg4l) [inline apiservers/portlayer/restapi/handlers/exec_handlers.go]: I'm guessing the 0th element in this list has some special significance.  Comment please.

File: apiservers/portlayer/restapi/handlers/exec_handlers.go [hunk 6]
```diff
@@ -122,9 +181,23 @@ func (handler *ExecHandlersImpl) ContainerCreateHandler(params exec.ContainerCre
...
 		},
+		Networks: make(map[string]metadata.NetworkEndpoint),
 	}
 	log.Infof("Metadata: %#v", m)
 
+	// network config
+	ns := params.CreateConfig.NetworkSettings
+	ne, s, err := addContainerToScope(name, ns)
+	defer func() {
+		if err != nil && s != nil {
```
Comment (fdawg4l) [inline apiservers/portlayer/restapi/handlers/exec_handlers.go]: This looks like a default error handler for this method.  You might want to log that an error is being caught here.

File: apiservers/portlayer/restapi/handlers/exec_handlers.go (hunk 3 above)
Comment (hmahmood) [inline apiservers/portlayer/restapi/handlers/exec_handlers.go]: docker's default network (i.e. bridge) vs. golang's default. I will add a comment.

File: apiservers/portlayer/restapi/handlers/exec_handlers.go (hunk 5 above)
Comment (hmahmood) [inline apiservers/portlayer/restapi/handlers/exec_handlers.go]: You guessed right :-) The call to Scopes() should return a slice with only one item, and preceding if checks for that. I will add a comment.

File: apiservers/portlayer/restapi/handlers/exec_handlers.go (hunk 4 above)
Comment (hmahmood) [inline apiservers/portlayer/restapi/handlers/exec_handlers.go]: Its a function call actually.

File: apiservers/portlayer/restapi/handlers/scopes_handlers.go [hunk 7]
```diff
@@ -30,7 +30,12 @@ import (
 
 // ScopesHandlersImpl is the receiver for all of the storage handler methods
 type ScopesHandlersImpl struct {
-	ctx *network.Context
+}
+
+var ctx *network.Context
```
Comment (fdawg4l) [inline apiservers/portlayer/restapi/handlers/scopes_handlers.go]: I know this is old code but consider renaming this.  We've been using `ctx` for `context.Context` everywhere else.

File: apiservers/portlayer/restapi/handlers/exec_handlers.go [hunk 8]
```diff
@@ -75,10 +77,69 @@ func (handler *ExecHandlersImpl) Configure(api *operations.PortLayerAPI) {
...
 
+func addContainerToScope(name string, ns *models.NetworkConfig) (*metadata.NetworkEndpoint, *network.Scope, error) {
+	if ns == nil {
+		return nil, nil, nil
+	}
+
+	var err error
+	var s *network.Scope
+	switch ns.NetworkName {
+	// docker's default network, usually maps to the defautl bridge network
```
Comment (caglar10ur) [inline apiservers/portlayer/restapi/handlers/exec_handlers.go]: typo: defautl

File: apiservers/portlayer/restapi/handlers/scopes_handlers.go [hunk 9]
```diff
@@ -30,7 +30,12 @@ import (
...
+
+func GetNetworkContext() *network.Context {
```
Comment (caglar10ur) [inline apiservers/portlayer/restapi/handlers/scopes_handlers.go]: this puts this ctx into package level and with a very confusing name. ctx used for context.Context extensively in vic repo. 

Why are we putting this variable into package level scope? If we really need to do that can we at least rename it to something else.

File: apiservers/portlayer/restapi/handlers/scopes_handlers.go [hunk 10]
```diff
@@ -30,7 +30,12 @@ import (
...
+	return ctx
 }
 
 // Configure assigns functions to all the storage api handlers
```
Comment (sflxn) [inline apiservers/portlayer/restapi/handlers/scopes_handlers.go]: cut-paste error.  "storage"

File: apiservers/portlayer/restapi/handlers/scopes_handlers.go (hunk 9 above)
Comment (hmahmood) [inline apiservers/portlayer/restapi/handlers/scopes_handlers.go]: I need to share it amongst all the handlers. Is there another way I can do that?

**Comment by caglar10ur** [comment]
LGTM :+1: :sheep: :it:

**Pull request merged**